catalog/catalog.go

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package catalog
+
+import (
+	"context"
+	"errors"
+
+	"github.com/apache/iceberg-go/table"
+	"github.com/aws/aws-sdk-go-v2/aws"
+)
+
+type CatalogType string
+
+const (
+	REST     CatalogType = "rest"
+	Hive     CatalogType = "hive"
+	Glue     CatalogType = "glue"
+	DynamoDB CatalogType = "dynamodb"
+	SQL      CatalogType = "sql"
+)
+
+var (
+	// ErrNoSuchTable is returned when a table does not exist in the catalog.
+	ErrNoSuchTable = errors.New("table does not exist")
+)
+
+// WithAwsConfig sets the AWS configuration for the catalog.
+func WithAwsConfig(cfg aws.Config) Option {
+	return func(o *Options) {
+		o.awsConfig = cfg
+	}
+}
+
+type Option func(*Options)
+
+type Options struct {
+	awsConfig aws.Config
+}
```

Comment:
Any reason to export these? We don't expect consumers to create their own `Option` functions right? If so we'd need to export the members of the `Options` struct which we don't.

Reply:
@zeroshade  Yeah I am a bit on the fence with this, feels half in and half out when the `Options` is private.

I am questioning whether this should be a Config structure with public members, but I wanted to get a couple more catalog implementations in and adjust it then.

Again keen to get a bit more code in the catalog and adjust it based on needs.

Reply:
If we're gonna go that route, I'd vote making the whole thing private for now, and then if we adjust it we can make it public or a public config struct later. 

Personally my default is always to make things private until it's explicitly necessary for it to be public. Keeps the surface API as small as possible

Reply:
@zeroshade cool, I have made options private so we can think about it once there is more implementations.